### ISSUE TITLE
Incorrect reverse operations for multiplication with zero-dimension tensors

### ISSUE THREAD

**Issue by VisualMelon** [DiffSharp/DiffSharp#131]
Products of scalars break reverse mode AD as either side of a multiplication. Example:

module DiffSharpTest.Program

```
open DiffSharp
open DiffSharp.Model

type ConstModel(size: int) =
    inherit Model()
    let b = Parameter <| dsharp.rand [size; size]
    do base.add([b], ["b"])
    override c.forward(value) =
        b.value
    member public this.B with get() = b

[<EntryPoint>]
let main argv =

    dsharp.config(backend=Backend.Reference)
    dsharp.seed(1)

    let model = ConstModel(10)
    model.reverseDiff()
    let loss = model.B.value.sum() * dsharp.tensor [1.0] // sum gives us a scalar: without it, everything works as you would expect
    loss.reverse()
    printfn "Grad: %A" model.B.value.derivative
    0
```

The gradients on `model.B` should be a matrix of ones, but instead it produces a scalar zero. This seems to be resolved by patching [`(*)`](https://github.com/DiffSharp/DiffSharp/blob/dev/src/DiffSharp.Core/Tensor.fs#L624-L643) as in https://github.com/VisualMelon/DiffSharp/commit/a3c599510a53d11d329e5dac85db6aa236384349, but I don't trust that I'm using the right 'overloads' still.

Other operations also look suspect in the `dim=0` case, e.g. [`(/)`](https://github.com/DiffSharp/DiffSharp/blob/dev/src/DiffSharp.Core/Tensor.fs#L680), but I'm not convinced I understand the T/T0 distinction properly yet, and come to a different conclusion every time I look at it, so I'm posting this before I end up in any more of a tangle. I'd be happy to do sort out a PR once I'm convinced I know what's going on.

These presumably will correspond to the currently missing tests, e.g. https://github.com/DiffSharp/DiffSharp/blob/dev/tests/DiffSharp.Tests/TestDerivatives.fs#L231-L234, which I'll look into writing

**Comment by gbaydin** [comment]
Thank you!
